mycore-restapi/src/main/java/org/mycore/restapi/v1/utils/MCRRestAPIUploadHelper.java

```diff
@@ -146,119 +145,172 @@ public static Response uploadObject(UriInfo info, HttpServletRequest request, In
-                            return categories.stream().allMatch(clazzSet::contains);
-                        }).findFirst();
-                    if (matchingDerivate.isPresent()) {
-                        derID = matchingDerivate.get().getXLinkHrefID();
-                    }
-                }
-            }
+            derID = findDerivateIDIfExists(mcrObj, label, classifications, overwriteOnExisting);
 
             if (derID == null) {
-                MCRDerivate mcrDerivate = new MCRDerivate();
-                if (label != null && label.length() > 0) {
-                    mcrDerivate.getDerivate().getTitles()
-                        .add(new MCRMetaLangText("title", null, null, 0, null, label));
-                }
-                MCRObjectID zeroDerId = MCRObjectID
-                    .getInstance(MCRObjectID.formatID(mcrObjIDObj.getProjectId() + "_derivate", 0));
-                mcrDerivate.setId(zeroDerId);
-                mcrDerivate.setSchema("datamodel-derivate.xsd");
-                mcrDerivate.getDerivate().setLinkMeta(new MCRMetaLinkID("linkmeta", mcrObjIDObj, null, null));
-                mcrDerivate.getDerivate().setInternals(new MCRMetaIFS("internal", null));
-
-                if (classifications != null && classifications.length() > 0) {
-                    final List<MCRMetaClassification> currentClassifications;
-                    currentClassifications = mcrDerivate.getDerivate().getClassifications();
-                    Stream.of(classifications.split(" "))
-                        .map(MCRCategoryID::fromString)
-                        .filter(dao::exist)
-                        .map(categoryID -> new MCRMetaClassification("classification", 0, null, categoryID))
-                        .forEach(currentClassifications::add);
-                }
-
-                MCRMetadataManager.create(mcrDerivate);
-                MCRMetadataManager.addOrUpdateDerivateToObject(mcrObjIDObj,
-                    MCRMetaEnrichedLinkIDFactory.getInstance().getDerivateLink(mcrDerivate),
-                    mcrDerivate.isImportMode());
-                derID = mcrDerivate.getId();
+                derID = createNewDerivate(mcrObjIDObj, label, classifications);
             }
 
-            response = Response
-                .created(info.getBaseUriBuilder().path("objects/" + mcrObjID + "/derivates/" + derID).build())
-                .type("application/xml; charset=UTF-8")
-                .build();
+            return buildResponse(info, mcrObjIDObj, String.valueOf(derID));
+
         } catch (Exception e) {
-            LOGGER.error("Exeption while uploading derivate", e);
+            LOGGER.error("Exception while uploading derivate", e);
+            return Response.status(Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    private static MCRObjectID findDerivateIDIfExists(MCRObject mcrObj, String label, String classifications,
+        boolean overwriteOnExisting) throws MCRRestAPIException {
```

Comment:
Please move `overwriteOnExisting` check to the parent method. Method name should be more like `findDerivateID`.